authn/saml/src/main/java/se/swedenconnect/signservice/authn/saml/SwedenConnectSamlAuthenticationHandler.java

```diff
@@ -225,6 +272,54 @@ protected boolean isSignMessageSupported() {
     return true;
   }
 
+  /**
+   * Checks if the IdP declares the entity category "http://id.elegnamnden.se/sprop/1.0/scal2".
+   */
+  @Override
+  protected boolean isSignatureActivationProtocolSupported(@Nonnull final EntityDescriptor idpMetadata) {
+
+    return EntityDescriptorUtils.getEntityCategories(idpMetadata).stream()
+        .filter(e -> EntityCategoryConstants.SERVICE_PROPERTY_CATEGORY_SCAL2.getUri().equals(e))
+        .findFirst()
+        .isPresent();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  protected void resetContext(@Nonnull final SignServiceContext context) {
+    super.resetContext(context);
+
+    context.remove(SAD_ID_KEY);
+  }
+
+  /**
+   * Checks for a SAD request.
+   */
+  @Override
+  protected void extendedAssertionVerification(@Nonnull final AuthnRequirements authnRequirements,
+      @Nonnull final AuthnRequest authnRequest, @Nonnull ResponseProcessingResult result,
+      @Nonnull final SignServiceContext context) throws UserAuthenticationException {
+
+    // Did we include a SADRequest extension in the AuthnRequest?
+    //
+    final String sadId = context.get(SAD_ID_KEY, String.class);
+    if (sadId == null) {
+      return;
+    }
+
+    // OK, we sent a SADRequest. We now expect the 'sad' attribute to be included in the assertion.
+    //
+    try {
+      this.sadValidator.validate(authnRequest, result.getAssertion());
+    }
```

Comment:
We are missing one check here (which is the reason why the doc count bug would go unnoticed).
This check checks that the SAD request is consistent with the SAD response.

What is missing is to check that the SAD request is consistent with what is actually being signed. I think the only check that remains is to verify just that the number of documents signed is equal to what is stated in the SAD request/response. In order to do this, I assume that we need to put the SAD data on the context and allow the sign handler to do this check.

Reply:
Yes. And now we set the doc count correctly.